[jax2tf] Clean up and expand binary elementwise harnesses.

bchetioui opened the pull request.

bchetioui committed:
[jax2tf] Clean up and expand binary elementwise harnesses.

gnecula commented on jax/experimental/jax2tf/tests/primitive_harness.py:
```diff
@@ -826,54 +812,141 @@ def _make_div_rem_harness(name, *, shapes=((2,), (2,)), dtype=np.float32,
-    (np.array([-1.6, -1.4, -1.0, 0.0, 0.1, 0.2, 1., 1.4, 1.6], dtype=dtype),
-     np.array([-1.6, 1.4, 1.0, 0.0, 0.1, 0.2, 1., 1.4, -1.6], dtype=dtype))
+) + tuple( # Validate broadcasting
+  _make_binary_elementwise_harness("broadcasting", prim=lax.atan2_p,
```
Is there a way to consolidate the broadcasting tests so we don't have to repeat this pattern?

bchetioui replied:
Hm I believe yes. This can probably be factorised a bit better – I maybe overly used `sed` :)
I think maybe `_make_binary_elementwise_harness` could become a function that only takes a primitive and a set of dtypes, and returns the whole tuple for both dtypes/broadcasting checks.
WDYT?

gnecula replied:
Sounds reasonable.

gnecula commented on jax/experimental/jax2tf/tests/primitive_harness.py:
```diff
@@ -826,54 +812,141 @@ def _make_div_rem_harness(name, *, shapes=((2,), (2,)), dtype=np.float32,
 )
 
-_LAX_BINARY_ELEMENTWISE = (
-  lax.add, lax.atan2, lax.div, lax.igamma, lax.igammac, lax.max, lax.min,
```
Are there all the binary ops tested? I did not see below the harnesses for add and and sub. I think that div/rem and min/max are already tested separately.

bchetioui replied:
Yes! `add` was already tested in `lax_add_mul` (line 721) and `sub` in `lax_sub` (line 516).

gnecula replied:
I see. Those don't have broadcasting. I wonder if it is important to add broadcasting checks everywhere. Maybe not.

bchetioui committed:
[jax2tf] Systematize broadcasting tests for binary elementwise harnesses.
Also add broadcasting tests to min and max, and splits the logic
for add and mul.

bchetioui commented:
This last commit should unify things the way we want. Note that I did not change the implementation of `lax_div_rem` because it requires paying some special attention to the generation of the arguments.

gnecula reviewed (approved):
Nice!
